eng/scripts/New-BuildInfo.ps1

```diff
@@ -0,0 +1,405 @@
+#!/bin/env pwsh
+#Requires -Version 7
+
+[CmdletBinding()]
+param(
+    [parameter(Mandatory)]
+    [ValidateSet('none', 'internal', 'public')]
+    [string] $PublishTarget,
+    [parameter(Mandatory)]
+    [int] $BuildId,
+    [string] $OutputPath,
+    [string] $ServerName,
+    [switch] $IncludeNative,
+    [switch] $TestPipeline
+)
+
+. "$PSScriptRoot/../common/scripts/common.ps1"
+. "$PSScriptRoot/helpers/PathHelpers.ps1"
+$RepoRoot = $RepoRoot.Path.Replace('\', '/')
+$isPipelineRun = $env:TF_BUILD -eq 'true'
+
+if(!$OutputPath) {
+    $OutputPath = "$RepoRoot/.work/build_info.json"
+}
+
+$serverDirectories = Get-ChildItem "$RepoRoot/servers" -Directory
+$toolDirectories = Get-ChildItem "$RepoRoot/tools" -Directory
+$coreDirectories = Get-ChildItem "$RepoRoot/core" -Directory
+
+$architectures = @('x64', 'arm64')
+
+$operatingSystems = @(
+    @{ name = 'linux'; nodeName = 'linux'; dotnetName = 'linux'; extension = '' }
+    @{ name = 'macos'; nodeName = 'darwin'; dotnetName = 'osx'; extension = '' }
+    @{ name = 'windows'; nodeName = 'win32'; dotnetName = 'win'; extension = '.exe' }
+)
+
+# We don't currently have pipeline support for building native on linux-arm64
+$excludedPlatforms = @('linux-arm64-native')
+
+# Public releases always use the version from the repo without a dynamic prerelease suffix, except for test pipelines
+# which always use a dynamic prerelease suffix to allow for multiple releases from the same commit
+$dynamicPrereleaseVersion = $PublishTarget -ne 'public' -or $TestPipeline
+
+function CheckVariable($name) {
+    $value = [Environment]::GetEnvironmentVariable($name)
+    if (-not $value) {
+        if ($isPipelineRun) {
+            Write-Error "Environment variable $name is not set."
+            exit 1
+        }
+        $substitute = "Missing-$name"
+        Write-Host "WARNING: Environment variable $name is not set. Using substitute value '$substitute'." -ForegroundColor Yellow
```

Comment:
We should probably use the logging.ps1 helper functions to these show up as warning in devops as well.